bin/scanpy_wrapper_utils.py

```diff
@@ -0,0 +1,156 @@
+
+
+    def add_subset_parameters(self):
+        self.parser.add_argument('-s', '--subset-names',
+                                 required=True,
+                                 type=comma_separated_list('subset-names', 'string'),
+                                 default=[],
+                                 help='Parameters to subset on.')
+        self.parser.add_argument('-l', '--low-thresholds',
+                                 type=comma_separated_list('low-thresholds', 'numeric'),
+                                 default=[],
+                                 help='Low cutoffs for the parameters (default is -Inf).')
+        self.parser.add_argument('-j', '--high-thresholds',
+                                 type=comma_separated_list('high-thresholds', 'numeric'),
+                                 default=[],
+                                 help='High cutoffs for the parameters (default is Inf).')
+        self.events.append({'handler':'_check_parameter_range',
+                            'argv':['subset_names','low_thresholds','high_thresholds']})
+
+
+    def _set_logging_level(self, argv):
+        debug = getattr(self.args, argv)
+        log_level = logging.DEBUG if debug else logging.WARN
+        logging.basicConfig(
+            level=log_level,
+            format='%(asctime)s; %(levelname)s; %(filename)s; %(funcName)s(): %(message)s',
+            datefmt='%y-%m-%d %H:%M:%S')
+
+
+    def _detect_io_format(self, argv):
+        fmt_key,fn_key = argv
+        fmt = getattr(self.args, fmt_key)
+        fn = getattr(self.args, fn_key)
+        if fmt != 'auto-detect':
+            return
+        if fn.endswith('.loom'):
+            setattr(self.args, fmt_key, 'loom')
+        elif fn.endswith('.h5ad'):
+            setattr(self.args, fmt_key, 'anndata')
+        else:
+            logging.error('Unspecified unknown format: "{}", '
+                          'please check suffix is either ".loom" or ".h5ad"'.format(fn))
+            sys.exit(1)
+
+
+    def _check_parameter_range(self, argv):
+        names,lows,highs = [getattr(self.args,k) for k in argv]
+        n = len(names)
+        if len(lows) == 0:
+            lows = [float('-Inf')] * n
+        elif len(lows) != n:
+            logging.error('--low-thresholds should be a comma separated list of numerics of the same size as {}'.format(argv[0]))
+            sys.exit(1)
+        if len(highs) == 0:
+            highs = [float('Inf')] * n
+        elif len(highs) != n:
+            logging.error('--high-thresholds should be a comma separated list of numerics of the same size as {}'.format(argv[0]))
+            sys.exit(1)
+
+
+    def _handle_event(self, ev):
+        handler = getattr(self, ev['handler'])
+        handler(ev['argv'])
+
+
+    def get_args(self):
+        self.args = self.parser.parse_args()
+
+        for ev in self.events:
+            self._handle_event(ev)
+
+        return self.args
+
+
+def read_input_object(filename, fmt):
```

Comment:
if we are not having documentation, I suggest being more explicit with argument names, like `format` instead of `fmt`.

Reply:
I've added documentation. pylint complains about "format" being a reserved word if used as variable name. I'll just leave it as "fmt" and hopefully documentation can help here.

Reply:
Sure, makes sense.